### ISSUE TITLE
Split file into multiple

### ISSUE THREAD

**Issue by Dolland** [xournalpp/xournalpp#4405]
Since i always produce 200+ pages files, saving takes a lot of time (10+ seconds).
So, besides the ability to turn saving a multithread job (i know there is some work under it) it can be nice to be able to split a large file into multiple files, just in case someone like me forget how saving became a long task as pages grown.
Thanks!

**Comment by eldipa** [comment]
@Dolland hi, the idea is to work on a new file format with a better performance than the current one (check #937 ). I will be publishing the design and progress but it is a long process however. 

If you can and want, you could submit you `.xopp` files for gathering statistics. Check #4862 to see what we learnt so far.

**Comment by Dolland** [comment]
> @Dolland hi, the idea is to work on a new file format with a better performance than the current one (check #937 ). I will be publishing the design and progress but it is a long process however.
> 
> If you can and want, you could submit you `.xopp` files for gathering statistics. Check #4862 to see what we learnt so far.

Hi!
Here are some files i made on xournal++
Hope they can help you!

https://drive.google.com/file/d/1rzP2csfM0Zoc_u1y72qdor3BRClzmAWT/view?usp=sharing

**Comment by eldipa** [comment]
Thanks! I took a look at your files and I understand why you wanted to split them into multiple ones and save them in background, I have a files of ~40MB too :)

**Comment by Dolland** [comment]
> Thanks! I took a look at your files and I understand why you wanted to split them into multiple ones and save them in background, I have a files of ~40MB too :)

well the digital electronics one is around 100 MB, it causes a massive delay when saving, so splitting could be a nice workaround while rewriting the saving system imho :)

**Comment by eldipa** [comment]
Yes, actually, it was my first idea as a workaround but I reviewed the code and I think it cannot be implemented directly.

Under the hood Xournal++ (mostly) does not know about the `.xopp` file: it has different objects in memory that represent all the elements of a document.

When a `.xopp` is read, it is read only once and all its content is loaded into memory. When you save the document, Xournal++ scans all the objects in memory and writes a single `.xopp` file.

Reading multiple `.xopp` files (like *parts* of the same document) it may be easy to implement (and it could be backward compatible). Writing may not.

Xournal++ does not know which objects (in memory) had changed or not so it always does a full scan and writes everything to disk. If we split the file into *parts* it will be have the same cost.

I know that implementing a new file format will *not* solve this: it should make the files smaller, yes, but Xournal++ will still write the file from scratch.

Making Xournal++ aware of what objects to store and which don't is *the key* to minimize the overhead. It is in my radar but it is a little far in the future.

**Comment by Dolland** [comment]
Uhm... I'm not aware of how xournal++ works, but I think we have different things in mind.

My suggestion was NOT to split a .xopp into multiple parts of the single .xopp in a manner that the file that i open remains one file but it's not "under the hood", but instead to generate multiple file by splitting, i.e., 100 pages into 4 .xopp made of 25 pages.

In this way you don't have anymore one file, but 4 file which pages are 0.25x the pages of the original file, like if you are exporting in a pdf and you select what pages must be exported.

I'm sorry if this is not clear and if i'm saying something stupid and undoable, but this was my idea.

**Comment by eldipa** [comment]
Actually is not a bad idea.

I coded a [python script](https://gist.github.com/eldipa/7dfff6bdb15e03dc5fe28c2df4eb3ee9) to may do exactly that.

For a given `uk-02.xopp` file, you can split it into multiple files of 25 pages each running:

```
python ~/proj/Code/xournal-file-format-xoz/dataset-xopp/split_xopp.py uk-02.xopp zzz 25
Counting pages...
uk-02.xopp has 430 pages
Writing pages 1 to 26 to zzz_001.xopp
Writing pages 26 to 51 to zzz_002.xopp
Writing pages 51 to 76 to zzz_003.xopp
...
Writing pages 401 to 426 to zzz_017.xopp
Writing pages 426 to 451 to zzz_018.xopp
```

The `zzz` is the prefix of the new created files.

Note that due some restriction of `.xopp`, the first page of your input file will be copied in all the output files.

I tested the script in one of my files and it works but I cannot guarantee that it will work in every case (`split_xopp.py` is a sort of a hack coded minutes ago)

**Comment by Dolland** [comment]
Really nice man! 
It could be implemented in a future nightly for who is interested, doesn't it?

**Comment by Dolland** [comment]
![immagine](https://github.com/xournalpp/xournalpp/assets/45694266/bdbe794c-f946-4206-835a-24549c97472f)

Actually i've just tried your script but i get the error in screenshot

**Comment by eldipa** [comment]
> ![immagine](https://user-images.githubusercontent.com/45694266/245075488-bdbe794c-f946-4206-835a-24549c97472f.png)
> 
> Actually i've just tried your script but i get the error in screenshot

@Dolland I think you are trying to execute a `.xopp` file instead of the python script. You are doing something like:
```
python foo.xopp split 25
```

What you have to do is download is this [python script](https://gist.github.com/eldipa/7dfff6bdb15e03dc5fe28c2df4eb3ee9) named `split_xopp.py`.

Download it into a folder (let's say folder `/home/user/Downloads`) and then run:

```
cd /home/user/Downloads
python split_xopp.py  foo.xopp zzz 25
```

Where `foo.xopp` is your `.xopp` file of interest (assuming that it is in the same folder)

**Comment by Dolland** [comment]
What a stupid i am, really sorry for this :⁠-⁠\

Il Sab 24 Giu 2023, 20:43 eldipa ***@***.***> ha scritto:
